### PR TITLE
IMPRO 827 CLI for temporal interpolation.

### PR DESCRIPTION
**Pull request by carolinejjones**
Description : Add command line interface for temporal interpolation

Testing:
 - [X] Ran tests and they passed OK
 - [X] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/utilities/temporal.py [hunk 1]
```diff
@@ -628,6 +629,7 @@ def process(self, cube_t0, cube_t1):
         interpolated_cube = cube.interpolate(time_list, iris.analysis.Linear())
         interpolated_cubes = iris.cube.CubeList()
         for single_time in interpolated_cube.slices_over('time'):
+            check_cube_not_float64(single_time, fix=True)
```
Comment (fionaRust) [inline lib/improver/utilities/temporal.py]: Why is this needed?

Comment (carolinejjones) [reply lib/improver/utilities/temporal.py]: Because iris.analysis converts forecast_period back to float64 and then sav_netcdf will not work as the cube is now float64.

File: bin/improver-temporal-interpolate [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        TemporalInterpolation(interval_in_minutes=interval,
+                              times=times).process(cube_t0,
+                                                   cube_t1))
+
+    len_files = len(args.output_files)
+    len_cubes = len(interpolated_cubes)
+    if len_files == len_cubes:
+        for i, cube_out in enumerate(interpolated_cubes):
+            save_netcdf(cube_out, args.output_files[i])
+    else:
```
[the remark this answers is not in the document]
Comment (carolinejjones) [reply bin/improver-temporal-interpolate]: Added

File: lib/improver/utilities/temporal.py [hunk 3]
```diff
@@ -627,10 +626,20 @@ def process(self, cube_t0, cube_t1):
...
+
+        # iris.analysis.Linear() modifies the dtype of time and forecast_period
+        # coords so need to revert back
+        dtype_time = cube_t0.coord('time').points.dtype
+        dtype_fp = cube_t0.coord('forecast_period').points.dtype
+
         interpolated_cubes = iris.cube.CubeList()
         for single_time in interpolated_cube.slices_over('time'):
-            check_cube_not_float64(single_time, fix=True)
+            coord_time = single_time.coord('time')
```
Comment (fionaRust) [inline lib/improver/utilities/temporal.py]: Do any unit tests need updating after this change?

Comment (carolinejjones) [reply lib/improver/utilities/temporal.py]: No because the unit tests are not making use of set_up_cube_variable at the moment. I was going to update these when I add the extra functionality to the interpolation code in the next PR.

File: bin/improver-temporal-interpolate [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    parser = ArgParser(
+        description='Interpolate data between validity times ')
+
+    parser.add_argument('data_t0', metavar='DATA_T0',
+                        help='Cube contain the data at the beginning'
+                        ' of the period')
+
+    parser.add_argument('data_t1', metavar='DATA_T1',
+                        help='Cube contain the data at the end'
+                        ' of the period')
```
Comment (cgsandford) [inline bin/improver-temporal-interpolate]: Could you rephrase these help strings to refer to (NetCDF) files not cubes please.

I'm also not quite happy with these variable names.  You could use "start_file" and "end_file", for example.  Alternatively you could have an "infiles" argument with "nargs=2" and sort the cubes by time inside the CLI.

File: lib/improver/utilities/temporal.py [hunk 5]
```diff
@@ -624,10 +625,21 @@ def process(self, cube_t0, cube_t1):
...
         cubes = iris.cube.CubeList([cube_t0, cube_t1])
-        cube = cubes.merge_cube()
+        cube = merge_cubes(cubes)
+
         interpolated_cube = cube.interpolate(time_list, iris.analysis.Linear())
+
+        # iris.analysis.Linear() modifies the dtype of time and forecast_period
+        # coords so need to revert back
+        dtype_time = cube_t0.coord('time').points.dtype
+        dtype_fp = cube_t0.coord('forecast_period').points.dtype
```
Comment (cgsandford) [inline lib/improver/utilities/temporal.py]: Are we sure the forecast period will be the only aux_coord ever associated with time?  It might be better if this could be more flexible - ie find any coords on the input cube that are associated with the time dimension, and cast them all back to their original datatypes.

File: bin/improver-temporal-interpolate [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                       "can not be set")
+
+    parser.add_argument("--output_files", metavar="OUTPUT_FILES",
+                        required=True, nargs="+",
+                        help="List of output files.")
+
+    args = parser.parse_args()
+
+    cube_t0 = load_cube(args.data_t0)
+    cube_t1 = load_cube(args.data_t1)
```
Comment (cgsandford) [inline bin/improver-temporal-interpolate]: It would add resilience to test for time here and reorder the cubes if needed.

File: bin/improver-temporal-interpolate [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    group = parser.add_mutually_exclusive_group(required=True)
+
+    group.add_argument("--interval_in_mins", metavar="INTERVAL_IN_MINS",
+                       default=None, type=int,
+                       help="Specifies the interval in minutes"
+                       " at which to interpolate "
+                       "between the two input cubes."
+                       " A number of minutes which does not "
+                       "divide up the interval equally will "
+                       "raise an exception. If intervals_in_mins "
```
Comment (cgsandford) [inline bin/improver-temporal-interpolate]: Could we have a CLI test for this exception please.

File: lib/improver/utilities/temporal.py (hunk 3 above)
Comment (cgsandford) [reply lib/improver/utilities/temporal.py]: I would also like to see a new test that coordinate datatype is unchanged.  But since the other work is happening in parallel, it can probably wait for that next PR.

File: bin/improver-temporal-interpolate (hunk 4 above)
Comment (carolinejjones) [reply bin/improver-temporal-interpolate]: I've replaced them with INFILES and check the order.

File: bin/improver-temporal-interpolate (hunk 6 above)
Comment (carolinejjones) [reply bin/improver-temporal-interpolate]: Done, The resulting interpolation cubes will still go from earliest to oldest

File: bin/improver-temporal-interpolate (hunk 7 above)
Comment (carolinejjones) [reply bin/improver-temporal-interpolate]: No there is already a unit test for this.

Comment (cgsandford) [reply bin/improver-temporal-interpolate]: I'm sorry but that's not my point.  What happens when somebody changes the plugin and its unit tests?  We have a comment here that says the CLI will behave in a specific way - presumably we need it to continue to behave in that way.  So we need a test for that please.

Comment (carolinejjones) [reply bin/improver-temporal-interpolate]: I will take out the comment if you like but I am not adding a CLI test. Sorry.

Comment (bayliffe) [reply bin/improver-temporal-interpolate]: As the exception is at the plugin level and is covered by a test we will forego a further test at the CLI level. We will leave the comment in as it is useful information to the user.

File: bin/improver-temporal-interpolate (hunk 4 above)
Comment (cgsandford) [reply bin/improver-temporal-interpolate]: Thanks.

File: lib/improver/utilities/temporal.py (hunk 5 above)
Comment (cgsandford) [reply lib/improver/utilities/temporal.py]: Fair enough, as long as it is captured.

File: bin/improver-temporal-interpolate (hunk 7 above)
Comment (cgsandford) [reply bin/improver-temporal-interpolate]: For the record, I am not happy with this.  If it is important for the CLI to behave in this way, then that behaviour needs to be covered by a CLI test.  The point of testing at all levels is to be able to identify wider impacts of any changes in the guts of the code, as we would not expect to have to be always aware of all the places we are calling a plugin or utility.  For the sake of 5 minutes of work, we are setting ourselves up for confusion and problems down the line.
